### PR TITLE
Online Reads Balancer: Fixing the online read balancer tests

### PR DESCRIPTION
**Pull request by pdhiran**
Fixed issues with the New Online reads balancer tests & other TFA issues.

### CODE REVIEW

**Comment by pdhiran** [comment]
Fail log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-CDX69Q 

Bug raised for PGs not scaling down upon bulk flag removal with upmap-read : https://bugzilla.redhat.com/show_bug.cgi?id=2302230


Current plan is to comment the code for scale-down events and check of scores post scale down.

**Comment by pdhiran** [comment]
Pass log without PG merge workflow : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-THOE3C

**Comment by pdhiran** [comment]
Run : CEPHQE-REEF-RADOS-REGRESSION-OPENSTACK-RH-18.2.1-229

Fail log : http://magna002.ceph.redhat.com/cephci-jenkins/results/openstack/RH/7.1/rhel-9/Regression/18.2.1-229/rados/165/tier-2_rados_test-bugfixes/Inconsistent_objects_in_replicated_pool_functionality_check_0.log

2024-08-01 13:51:28,836 (cephci.test_osd_replicated_inconsistency_scenario) [DEBUG] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.ceph.ceph_admin.shell.py:58 - HEALTH_ERR 8 scrub errors; Possible data damage: 2 pgs inconsistent
[ERR] OSD_SCRUB_ERRORS: 8 scrub errors
[ERR] PG_DAMAGED: Possible data damage: 2 pgs inconsistent
pg 11.8 is active+clean+inconsistent, acting [8,3,7]
pg 11.18 is active+clean+inconsistent, acting [13,9,5]

2024-08-01 13:51:29,279 (cephci.test_osd_replicated_inconsistency_scenario) [INFO] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.tests.rados.test_osd_replicated_inconsistency_scenario.py:261 - The 4 objects are converted into inconsistent objects
2024-08-01 13:51:29,280 (cephci.test_osd_replicated_inconsistency_scenario) [INFO] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.tests.rados.test_osd_replicated_inconsistency_scenario.py:263 - The inconsistent count on the replicated pool is - 4

Pass log post fix : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-23OBL0

**Comment by pdhiran** [comment]
RUN : CEPHQE-REEF-RADOS-REGRESSION-OPENSTACK-RH-18.2.1-229 

Fail log : http://magna002.ceph.redhat.com/cephci-jenkins/results/openstack/RH/7.1/rhel-9/Regression/18.2.1-229/rados/165/tier-2_rados_test-bugfixes/Verification_the_recovery_flag_functionality_0.log

snippet : 
2024-08-01 18:28:12,572 (cephci.test_norecovery_flag_functionality) [DEBUG] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.utility.utils.py:1858 - The <function wait_for_clean_pg_sets at 0x7ff659e50ee0> return status is False
2024-08-01 18:28:12,572 (cephci.test_norecovery_flag_functionality) [INFO] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.tests.rados.test_norecovery_flag_functionality.py:265 - Execution failed at function <function wait_for_clean_pg_sets at 0x7ff659e50ee0>
2024-08-01 18:28:12,573 (cephci.test_norecovery_flag_functionality) [INFO] - cephci.RH.7.1.rhel-9.Regression.18.2.1-229.rados.165.cephci.tests.rados.test_norecovery_flag_functionality.py:266 - Traceback (most recent call last):
  File "/home/jenkins/ceph-builds/openstack/RH/7.1/rhel-9/Regression/18.2.1-229/rados/165/cephci/tests/rados/test_norecovery_flag_functionality.py", line 262, in run
    delete_and_healthCheck(rados_obj, pool_name)
  File "/home/jenkins/ceph-builds/openstack/RH/7.1/rhel-9/Regression/18.2.1-229/rados/165/cephci/tests/rados/test_norecovery_flag_functionality.py", line 315, in delete_and_healthCheck
    method_should_succeed(wait_for_clean_pg_sets, rados_obj)
  File "/home/jenkins/ceph-builds/openstack/RH/7.1/rhel-9/Regression/18.2.1-229/rados/165/cephci/utility/utils.py", line 1860, in method_should_succeed
    raise AssertionError(f"Execution failed at function {function}")
AssertionError: Execution failed at function <function wait_for_clean_pg_sets at 0x7ff659e50ee0>


Fail log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-4E82BB
BZ comment : https://bugzilla.redhat.com/show_bug.cgi?id=2134786#c48 

Pass with commented manual PG change : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-NM01N7

File: tests/rados/test_norecovery_flag_functionality.py [hunk 1]
```diff
@@ -38,195 +38,288 @@ def run(ceph_cluster, **kw):
...
-            f"The pg number is -{pg_num}"
+            "Scenario2:The pg number is decreased after unsetting the norecover"
+            " flag upon bulk flag addition/ removal complete"
         )
         delete_and_healthCheck(rados_obj, pool_name)
         log.info("=========Scenario2: Test completed===========")
+        time.sleep(20)
+
         log.info(
             "Scenario3:Modify the pg number manually and verify that the pg_num should not increase "
```
Comment (SrinivasaBharath) [inline tests/rados/test_norecovery_flag_functionality.py:195]: Following are the manual steps of scenario-3 and 4 which are automated -
ceph osd set norecover
ceph osd pool set  test pg_num 8
Check the “ceph osd pool autoscale-status” .The pg_num should not be increased
ceph osd unset norecover
**Verification step**- The pg_num should be increased. 
ceph osd pool set  test pg_num 512
ceph osd set norecover
rados bench -p test 5 write -b 1024 --no-cleanup
**Verification Step** -Check the “ceph osd pool autoscale-status” .The pg_num should not be decreased
ceph osd unset norecover

Modified Code is(From bugzilla) -
1. Set norecovery flag on the cluster
 ceph osd set norecover
2. Create pool with 8 PGs
   ceph osd pool create test 8
3. As expected, the autoscaling of the PGs is not happening.
4. Change the PG count on the pool. See that PGs start scaling. 
cmd : ceph osd pool set <pool> pg_num 256

We can add this as a new scenario.

**Review by SrinivasaBharath — CHANGES_REQUESTED** [review]

File: tests/rados/test_norecovery_flag_functionality.py (hunk 1 above)
Comment (pdhiran) [reply tests/rados/test_norecovery_flag_functionality.py:195]: > Following are the manual steps of scenario-3 and 4 which are automated - ceph osd set norecover ceph osd pool set test pg_num 8 Check the “ceph osd pool autoscale-status” .The pg_num should not be increased ceph osd unset norecover 
-> this is already covered as part of scenario 1.

**Verification step**- The pg_num should be increased. ceph osd pool set test pg_num 512 ceph osd set norecover rados bench -p test 5 write -b 1024 --no-cleanup 
**Verification Step** -Check the “ceph osd pool autoscale-status” .The pg_num should not be decreased ceph osd unset norecover
-> this is already covered as part of scenario 3, Which is asked about in the bugzilla. 

> 
> Modified Code is(From bugzilla) -
> 
> 1. Set norecovery flag on the cluster
>    ceph osd set norecover
> 2. Create pool with 8 PGs
>    ceph osd pool create test 8
> 3. As expected, the autoscaling of the PGs is not happening.
> 4. Change the PG count on the pool. See that PGs start scaling.
>    cmd : ceph osd pool set  pg_num 256
> 
> We can add this as a new scenario.

**Review by SrinivasaBharath — APPROVED** [review]
looks good to me.

**Review by neha-gangadhar — APPROVED** [review]
Approved

**Pull request merged**